src/LayerEditor/ui/data/block_item.py

```diff
@@ -0,0 +1,142 @@
+
+
+    @property
+    def decomposition(self):
+        # This is going to be problem in future when there will be more editable layers in one block
+        for layer in self.layers:
+            if isinstance(layer.top_top, InterfaceNodeSetItem):
+                return layer.top_top.decomposition
+
+    @property
+    def layer_names(self):
+        for layer in self.layers:
+            yield layer.name
+
+    def make_node_set_from_data(self, le_model, node_set: [InterfaceNodeSet, InterpolatedNodeSet]):
+        if isinstance(node_set, InterfaceNodeSet):
+            le_model.decompositions_model.decomps[node_set.nodeset_id].helper_attr_block = self
+            # I don't have any other sensible idea how to keep track of topology_id inside decomposition
+            return InterfaceNodeSetItem(le_model.decompositions_model.decomps[node_set.nodeset_id],
+                                        le_model.interfaces_model.interfaces[node_set.interface_id])
+        else:
+            interfaces = le_model.interfaces_model.interfaces
+            decomp = le_model.decompositions_model.decomps[node_set.surf_nodesets[0].nodeset_id]
+            itf = interfaces[node_set.surf_nodesets[0].interface_id]
+            itf_node_set1 = InterfaceNodeSetItem(decomp, itf)
+            decomp = le_model.decompositions_model.decomps[node_set.surf_nodesets[1].nodeset_id]
+            itf = interfaces[node_set.surf_nodesets[1].interface_id]
+            itf_node_set2 = InterfaceNodeSetItem(decomp, itf)
+
+            itf = interfaces[node_set.interface_id]
+
+            return InterpolatedNodeSetItem(itf_node_set1, itf_node_set2, itf)
+
+
+    def init_add_layer(self, layer_data: [StratumLayer, FractureLayer], le_model):
+        """Add layer while initializing (isn't undoable)."""
+        top_top = self.make_node_set_from_data(le_model, layer_data.top)
+        if hasattr(layer_data, "bottom"):
+            bottom_top = self.make_node_set_from_data(le_model, layer_data.bottom)
+        else:
+            bottom_top = None
+
+        shape_regions = [{}, {}, {}]
+
+        for shape_id, region_id in enumerate(layer_data.node_region_ids):
+            shape_regions[0][shape_id] = self.regions_model.regions.get(region_id)
+
+        for shape_id, region_id in enumerate(layer_data.segment_region_ids):
+            shape_regions[1][shape_id] = self.regions_model.regions.get(region_id)
+
+        for shape_id, region_id in enumerate(layer_data.polygon_region_ids):
+            shape_regions[2][shape_id] = self.regions_model.regions.get(region_id)
+
+        layer = LayerItem(self,
+                          layer_data.name,
+                          top_top,
+                          bottom_top,
+                          shape_regions)
+        self.layers.append(layer)
+        self.layers_dict.add(layer)
+
+        return layer
+
+    def init_regions_for_new_shape(self, shape: [GsPoint, GsSegment, GsPolygon]):
```

Comment:
Where is this used? What is special about setting the region to a new shape.

Reply:
It is used when new shape is created. It assigns default (not always None) region to the new shape for every layer where the shape exists.